File: internal/httpapi/market_test.go
package httpapi

import (
	"context"
	"database/sql"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/DeliciousBuding/fund-dashboard/internal/config"
	"github.com/DeliciousBuding/fund-dashboard/internal/testutil"
)

func TestMarketAndStockRoutesExposeCachedFactsOnlyData(t *testing.T) {
	db := openMarketHTTPFixture(t)
	defer db.Close()

	router := newAuthedRouter(t, config.Config{ServiceName: "fund-dashboard-go", Version: "test"}, db)

	// REST contract is MarketIndex[] for the SPA; MCP keeps the service envelope.
	indices := doJSONArrayRequest(t, router, http.MethodGet, "/api/market/indices", http.StatusOK)
	if len(indices) != 2 {
		t.Fatalf("indices len = %d, want 2; body=%s", len(indices), toJSONString(t, indices))
	}
	byCode := map[string]map[string]any{}
	for _, row := range indices {
		code, _ := row["code"].(string)
		byCode[code] = row
	}
	gspc := byCode["^GSPC"]
	if gspc == nil || gspc["name"] != "标普500" || gspc["price"].(float64) != 5600.5 {
		t.Fatalf("GSPC = %#v, want cached row", gspc)
	}
	if gspc["change_amt"].(float64) != 23.5 || gspc["change_pct"].(float64) != 0.42 {
		t.Fatalf("GSPC change fields = %#v", gspc)
	}

	stock := doJSONRequest(t, router, http.MethodGet, "/api/stocks/aapl?range=1y&include_history=true", nil, http.StatusOK)
	// SPA contract is flat USStockInfo (#98).
	if stock["code"] != "AAPL" ||
		stock["decision_boundary"] != "facts_only" ||
		stock["side_effects"] != "none" ||
		stock["external_fetch"] != "not_performed" {
		t.Fatalf("stock response = %s", toJSONString(t, stock))
	}
	if stock["name"] != "Apple Inc." || stock["price"].(float64) != 198.25 {
		t.Fatalf("quote fields = %s", toJSONString(t, stock))
	}
	history, ok := stock["history"].([]any)
	if !ok || len(history) != 2 {
		t.Fatalf("history = %#v, want 2 points", stock["history"])
	}
	first := history[0].(map[string]any)
	if first["date"] != "2026-06-18" && first["date"] != "2026-06-17" {
		// order is newest-first from service
		if first["close"] == nil {
			t.Fatalf("history point = %#v", first)
		}
	}
	if strings.Contains(toJSONString(t, stock), "backup") ||
		strings.Contains(toJSONString(t, stock), "建议买入") ||
		strings.Contains(toJSONString(t, stock), "external_fetch_performed") {
		t.Fatalf("stock route should stay facts-only: %s", toJSONString(t, stock))
	}
}

func TestExchangeRateRouteFetchesConfiguredFactsOnlySource(t *testing.T) {
	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{
			"chart": {
				"result": [{
					"meta": {
						"regularMarketPrice": 7.2513,
						"regularMarketTime": 1783439100
					}
				}]
			}
		}`))
	}))
	defer upstream.Close()
	t.Setenv("FUND_EXCHANGE_RATE_ENDPOINT", upstream.URL)

	db := openMarketHTTPFixture(t)
	defer db.Close()

	router := newAuthedRouter(t, config.Config{ServiceName: "fund-dashboard-go", Version: "test"}, db)
	rate := doJSONRequest(t, router, http.MethodGet, "/api/market/exchange-rate", nil, http.StatusOK)
	if rate["from"] != "USD" ||
		rate["to"] != "CNY" ||
		rate["rate"].(float64) != 7.2513 ||
		rate["updated_at"] == "" {
		t.Fatalf("exchange rate response = %s", toJSONString(t, rate))
	}
	if strings.Contains(toJSONString(t, rate), "backup") ||
		strings.Contains(toJSONString(t, rate), "建议买入") {
		t.Fatalf("exchange rate route should stay facts-only: %s", toJSONString(t, rate))
	}
}

func TestIndexLiveAndHistoryRoutes(t *testing.T) {
	// Stub Yahoo history via service package hooks is package-private; use live fixture indices for live quote.
	db := openMarketHTTPFixture(t)
	defer db.Close()
	router := newAuthedRouter(t, config.Config{ServiceName: "fund-dashboard-go", Version: "test"}, db)

	// Live quote from indices cache (fresh fixture timestamps from #92).
	live := doJSONRequest(t, router, http.MethodGet, "/api/market/index/GSPC", nil, http.StatusOK)
	if live["code"] != "^GSPC" && live["code"] != "GSPC" {
		// Normalize yields ^GSPC
		if live["code"] != "^GSPC" {
			// accept either if handler returns raw — should be ^GSPC
			if live["price"] == nil {
				t.Fatalf("live = %s", toJSONString(t, live))
			}
		}
	}
	if live["price"].(float64) != 5600.5 {
		// may refresh from yahoo if fixture stale; at least ensure route works
		if live["error"] != nil && live["error"] != "" {
			t.Fatalf("live error = %s", toJSONString(t, live))
		}
	}

	// History: may hit real Yahoo in CI if no stub — skip hard fail on network.
	// Instead unit-test history via portfolio package; here only ensure route exists (not 404).
	req := httptest.NewRequest(http.MethodGet, "/api/market/index/NDX/history?range=1y", nil)
	rec := httptest.NewRecorder()
	router.ServeHTTP(rec, req)
	if rec.Code == http.StatusNotFound {
		t.Fatalf("history route 404 — route not registered")
	}
	if rec.Code != http.StatusOK && rec.Code != http.StatusBadGateway {
		t.Fatalf("history status = %d body=%s", rec.Code, rec.Body.String())
	}
}

func openMarketHTTPFixture(t *testing.T) *sql.DB {
	t.Helper()
	return testutil.OpenTempDBWithSchema(t, marketHTTPFixtureStatements)
}

var marketHTTPFixtureStatements = []string{
	`CREATE TABLE indices (
		code TEXT PRIMARY KEY,
		name TEXT,
		market TEXT,
		price REAL,
		change_pct REAL,
		change_amt REAL,
		updated_at TEXT DEFAULT (datetime('now'))
	)`,
	`CREATE TABLE stock_realtime (
		code TEXT,
		market TEXT,
		name TEXT,
		price REAL,
		open REAL,
		high REAL,
		low REAL,
		change_pct REAL,
		change_amt REAL,
		volume REAL,
		amount REAL,
		turnover REAL,
		pe REAL,
		total_mv REAL,
		circ_mv REAL,
		high52 REAL,
		low52 REAL,
		currency TEXT DEFAULT '',
		updated_at TEXT DEFAULT (datetime('now')),
		PRIMARY KEY (code, market)
	)`,
	`CREATE TABLE stock_kline_cache (
		code TEXT,
		market TEXT,
		date TEXT,
		open REAL,
		close REAL,
		high REAL,
		low REAL,
		volume REAL,
		amount REAL,
		amplitude REAL,
		change_pct REAL,
		turnover_rate REAL,
		PRIMARY KEY (code, market, date)
	)`,
	`CREATE TABLE stock_profile (
		code TEXT,
		name TEXT,
		market TEXT,
		sector TEXT,
		industry TEXT,
		market_cap REAL,
		pe REAL,
		description TEXT,
		PRIMARY KEY (code, market)
	)`,
	`INSERT INTO indices (code, name, market, price, change_pct, change_amt, updated_at) VALUES
		('^GSPC', '标普500', 'US', 5600.5, 0.42, 23.5, '2099-01-01 12:00:00'),
		('^NDX', '纳斯达克100', 'US', 19888.2, 1.25, 245.8, '2099-01-01 12:00:00')`,
	`INSERT INTO stock_realtime (code, market, name, price, open, high, low, change_pct, change_amt, volume, amount, pe, total_mv, high52, low52, currency, updated_at)
		VALUES ('AAPL', 'US', 'Apple Inc.', 198.25, 196.5, 199.0, 195.8, 1.2, 2.35, 45000000, 8900000000, 31.2, 3000000000000, 205.0, 160.0, 'USD', '2099-01-01 12:00:00')`,
	`INSERT INTO stock_kline_cache (code, market, date, open, close, high, low, volume, change_pct) VALUES
		('AAPL', 'US', '2026-06-18', 196.5, 198.25, 199.0, 195.8, 45000000, 1.2),
		('AAPL', 'US', '2026-06-17', 194.0, 195.9, 196.2, 193.7, 41000000, 0.7)`,
	`INSERT INTO stock_profile (code, name, market, sector, industry, market_cap, pe, description)
		VALUES ('AAPL', 'Apple Inc.', 'US', 'Technology', 'Consumer Electronics', 3000000000000, 31.2, 'Consumer hardware and services')`,
}

func TestMarketStreamEmitsIndicesEvent(t *testing.T) {
	db := openMarketHTTPFixture(t)
	defer db.Close()
	router := newAuthedRouter(t, config.Config{ServiceName: "fund-dashboard-go", Version: "test"}, db)

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	req := httptest.NewRequestWithContext(ctx, http.MethodGet, "/api/market/stream", nil)
	rec := httptest.NewRecorder()

	done := make(chan struct{})
	go func() {
		defer close(done)
		router.ServeHTTP(rec, req)
	}()

	// Wait briefly for first frame then cancel, then block until the handler
	// actually returns so reading rec.Body below does not race the SSE goroutine.
	time.Sleep(200 * time.Millisecond)
	cancel()
	select {
	case <-done:
	case <-time.After(2 * time.Second):
		t.Fatal("market stream handler did not return after cancel")
	}

	if rec.Code != http.StatusOK {
		// httptest may still be 200 once headers written
	}
	ct := rec.Header().Get("Content-Type")
	if !strings.Contains(ct, "text/event-stream") {
		t.Fatalf("Content-Type = %q, want text/event-stream; body=%s", ct, rec.Body.String()[:min(200, rec.Body.Len())])
	}
	body := rec.Body.String()
	if !strings.Contains(body, "event: indices") {
		t.Fatalf("body missing indices event: %s", body[:min(400, len(body))])
	}
	if !strings.Contains(body, "^GSPC") && !strings.Contains(body, "GSPC") {
		t.Fatalf("body missing index code: %s", body[:min(400, len(body))])
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: packages/contracts/schemas/market.test.ts
// market.test.ts — /api/market/* wire-shape tests (node:test).
// 对照 internal/httpapi/market.go 与 internal/service/portfolio/index_history.go：
// indices 7 键恒下发；history 8 键 map 字面量恒下发（error/message 不进该 map）；
// live 7 恒发键 + 6 omitempty 条件键。
import assert from "node:assert/strict";
import { test } from "node:test";

import {
  ExchangeRateSchema,
  IndexHistoryPointSchema,
  IndexHistorySchema,
  IndexLiveSchema,
  MarketIndexSchema,
} from "./market.ts";

const marketIndexWire = {
  code: "^GSPC",
  name: "标普500",
  market: "US",
  price: 5600.5,
  change_pct: 0.42,
  change_amt: 23.5,
  updated_at: "2099-01-01 12:00:00",
};

const indexHistoryPointWire = { date: "2026-08-28", close: 19888.2, change_pct: 1.25 };

const indexHistoryWire = {
  symbol: "^NDX",
  count: 2,
  range: "1y",
  data: [indexHistoryPointWire, { date: "2026-08-29", close: 19910.5, change_pct: 0.11 }],
  source: "yahoo_chart",
  external_fetch: "yahoo_chart",
  decision_boundary: "facts_only",
  side_effects: "none",
};

const indexLiveWire = {
  code: "^GSPC",
  name: "标普500",
  market: "US",
  price: 5600.5,
  change_pct: 0.42,
  change_amt: 23.5,
  updated_at: "2099-01-01 12:00:00",
  source: "indices_cache",
  decision_boundary: "facts_only",
  side_effects: "none",
  external_fetch: "not_performed",
};

const exchangeRateWire = {
  from: "USD",
  to: "CNY",
  rate: 7.2345,
  updated_at: "2026-08-29T09:30:00Z",
  source: "yahoo_chart",
};

// ── /api/market/indices ──────────────────────────────────────────────

test("MarketIndexSchema parses the real indices shape", () => {
  const parsed = MarketIndexSchema.parse(marketIndexWire);
  assert.equal(parsed.code, "^GSPC");
  assert.equal(parsed.price, 5600.5);
  assert.equal(parsed.change_pct, 0.42);
});

test("MarketIndexSchema rejects null price (Go float64 defaults to 0)", () => {
  assert.throws(() => MarketIndexSchema.parse({ ...marketIndexWire, price: null }));
});

test("MarketIndexSchema requires updated_at (Go always emits it)", () => {
  const { updated_at, ...rest } = marketIndexWire;
  assert.throws(() => MarketIndexSchema.parse(rest));
});

test("MarketIndexSchema rejects unknown fields (live keys do not belong here)", () => {
  assert.throws(() => MarketIndexSchema.parse({ ...marketIndexWire, source: "yahoo_chart" }));
});

// ── /api/market/index/{code}/history ─────────────────────────────────

test("IndexHistorySchema parses the real 8-key shape", () => {
  const parsed = IndexHistorySchema.parse(indexHistoryWire);
  assert.equal(parsed.symbol, "^NDX");
  assert.equal(parsed.count, 2);
  assert.equal(parsed.source, "yahoo_chart");
  assert.equal(parsed.decision_boundary, "facts_only");
  assert.equal(parsed.data.length, 2);
});

test("IndexHistorySchema requires source (map literal always emits it)", () => {
  const { source, ...rest } = indexHistoryWire;
  assert.throws(() => IndexHistorySchema.parse(rest));
});

test("IndexHistorySchema requires external_fetch (map literal always emits it)", () => {
  const { external_fetch, ...rest } = indexHistoryWire;
  assert.throws(() => IndexHistorySchema.parse(rest));
});

test("IndexHistorySchema requires decision_boundary (map literal always emits it)", () => {
  const { decision_boundary, ...rest } = indexHistoryWire;
  assert.throws(() => IndexHistorySchema.parse(rest));
});

test("IndexHistorySchema requires side_effects (map literal always emits it)", () => {
  const { side_effects, ...rest } = indexHistoryWire;
  assert.throws(() => IndexHistorySchema.parse(rest));
});

test("IndexHistorySchema requires data (always emitted)", () => {
  const { data, ...rest } = indexHistoryWire;
  assert.throws(() => IndexHistorySchema.parse(rest));
});

test("IndexHistorySchema rejects unknown fields (contract drift)", () => {
  assert.throws(() => IndexHistorySchema.parse({ ...indexHistoryWire, future_field: "surprise" }));
});

test("IndexHistoryPointSchema parses the real point shape", () => {
  const parsed = IndexHistoryPointSchema.parse(indexHistoryPointWire);
  assert.equal(parsed.close, 19888.2);
});

test("IndexHistoryPointSchema rejects a point missing close", () => {
  const { close, ...rest } = indexHistoryPointWire;
  assert.throws(() => IndexHistoryPointSchema.parse(rest));
});

test("IndexHistoryPointSchema rejects unknown fields", () => {
  assert.throws(() =>
    IndexHistoryPointSchema.parse({ ...indexHistoryPointWire, future_field: "surprise" }),
  );
});

// ── /api/market/index/{code}（live）──────────────────────────────────

test("IndexLiveSchema parses the real indices-cache shape", () => {
  const parsed = IndexLiveSchema.parse(indexLiveWire);
  assert.equal(parsed.code, "^GSPC");
  assert.equal(parsed.price, 5600.5);
  assert.equal(parsed.source, "indices_cache");
  assert.equal(parsed.error, undefined);
});

test("IndexLiveSchema tolerates omitempty source being absent", () => {
  const { source, ...rest } = indexLiveWire;
  const parsed = IndexLiveSchema.parse(rest);
  assert.equal(parsed.source, undefined);
});

test("IndexLiveSchema tolerates omitempty error/message being absent", () => {
  const parsed = IndexLiveSchema.parse(indexLiveWire);
  assert.equal(parsed.error, undefined);
  assert.equal(parsed.message, undefined);
});

test("IndexLiveSchema parses the no-data shape (error/message, no source)", () => {
  const parsed = IndexLiveSchema.parse({
    code: "^ZZZ",
    name: "",
    market: "",
    price: 0,
    change_pct: 0,
    change_amt: 0,
    updated_at: "",
    decision_boundary: "facts_only",
    side_effects: "none",
    external_fetch: "not_performed",
    error: "no_data",
    message: "no quote for ^ZZZ",
  });
  assert.equal(parsed.error, "no_data");
  assert.equal(parsed.message, "no quote for ^ZZZ");
  assert.equal(parsed.source, undefined);
});

test("IndexLiveSchema requires code (Go always emits it)", () => {
  const { code, ...rest } = indexLiveWire;
  assert.throws(() => IndexLiveSchema.parse(rest));
});

test("IndexLiveSchema requires updated_at (Go always emits it, even empty)", () => {
  const { updated_at, ...rest } = indexLiveWire;
  assert.throws(() => IndexLiveSchema.parse(rest));
});

test("IndexLiveSchema rejects unknown fields (contract drift)", () => {
  assert.throws(() => IndexLiveSchema.parse({ ...indexLiveWire, future_field: "surprise" }));
});

// ── /api/market/exchange-rate ────────────────────────────────────────

test("ExchangeRateSchema parses the real report shape", () => {
  const parsed = ExchangeRateSchema.parse(exchangeRateWire);
  assert.equal(parsed.from, "USD");
  assert.equal(parsed.to, "CNY");
  assert.equal(parsed.rate, 7.2345);
  assert.equal(parsed.source, "yahoo_chart");
});

test("ExchangeRateSchema tolerates omitempty source being absent", () => {
  const { source, ...withoutSource } = exchangeRateWire;
  const parsed = ExchangeRateSchema.parse(withoutSource);
  assert.equal(parsed.source, undefined);
  assert.equal(parsed.rate, 7.2345);
});

test("ExchangeRateSchema rejects a report missing rate", () => {
  const { rate, ...rest } = exchangeRateWire;
  assert.throws(() => ExchangeRateSchema.parse(rest));
});

test("ExchangeRateSchema rejects unknown fields (contract drift)", () => {
  assert.throws(() => ExchangeRateSchema.parse({ ...exchangeRateWire, future_field: "surprise" }));
});
